prow/plugins/dco/dco.go

```diff
@@ -0,0 +1,356 @@
+		// prevent us commenting every time a label is added/deleted.
+		cp.PruneComments(shouldPrune(l, botName))
+		l.Infof("Commenting on PR to advise users of DCO check")
+		if err := gc.CreateComment(org, repo, pr.Number, fmt.Sprintf(dcoNotFoundMessage, targetURL, markdownSHAList(org, repo, commitsMissingDCO), plugins.AboutThisBot)); err != nil {
+			l.WithError(err).Warning("Could not create DCO not found comment.")
+		}
+	}
+
+	return nil
+}
+
+// 1. Check commit messages in the pull request for the sign-off string
+// 2. Check the existing status context value
+// 3. Check the existing PR labels
+// 4. If signed off, apply appropriate labels and status context.
+// 5. If not signed off, apply appropriate labels and status context and add a comment.
+func handle(gc gitHubClient, cp commentPruner, log *logrus.Entry, org, repo string, pr github.PullRequest, addComment bool) error {
+	l := log.WithField("pr", pr.Number)
+
+	commitsMissingDCO, err := checkCommitMessages(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error running DCO check against commits in PR")
+		return err
+	}
+
+	existingStatus, err := checkExistingStatus(gc, l, org, repo, pr.Head.SHA)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR status")
+		return err
+	}
+
+	hasYesLabel, hasNoLabel, err := checkExistingLabels(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR labels")
+		return err
+	}
+
+	return takeAction(gc, cp, l, org, repo, pr, commitsMissingDCO, existingStatus, hasYesLabel, hasNoLabel, addComment)
+}
+
+func markdownSHAList(org, repo string, list []github.GitCommit) string {
+	lines := make([]string, len(list))
+	lineFmt := "- [%s](https://github.com/%s/%s/commits/%s) %s"
+	for i, commit := range list {
+		if commit.SHA == "" {
+			continue
+		}
+		// if we somehow encounter a SHA that's less than 7 characters, we will
+		// just use it as is.
+		shortSHA := commit.SHA
+		if len(shortSHA) > 7 {
+			shortSHA = shortSHA[:7]
+		}
+
+		// get the first line of the commit
+		message := strings.Split(commit.Message, "\n")[0]
+
+		lines[i] = fmt.Sprintf(lineFmt, shortSHA, org, repo, commit.SHA, message)
+	}
+	return strings.Join(lines, "\n")
+}
+
+// shouldPrune finds comments left by this plugin.
+func shouldPrune(log *logrus.Entry, botName string) func(github.IssueComment) bool {
+	return func(comment github.IssueComment) bool {
+		return strings.Contains(comment.Body, dcoMsgPruneMatch)
+	}
+}
+
+func handlePullRequestEvent(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	return handlePullRequest(pc.GitHubClient, pc.CommentPruner, pc.Logger, pe)
+}
+
+func handlePullRequest(gc gitHubClient, cp commentPruner, log *logrus.Entry, pe github.PullRequestEvent) error {
+	org := pe.Repo.Owner.Login
+	repo := pe.Repo.Name
+
+	// we only reprocess on label, unlabel, open, reopen and synchronize events
```

Comment:
> If so, usually it is better to allow users with labeling permissions to change the labels in case of emergency

Makes sense to me. I'll update the PR to not handle label/unlabel events.